nginx-unit 1.1 (new formula)

sj26 opened the pull request:
Add a formula for NGINX's new [Unit](https://unit.nginx.org). I debated the name — their project is called "unit" and it produces a binary `unitd`, but maybe the formula should be `nginx-unit`? Unsure.

---

- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

sj26 commented:
Um, could it not be closed?

commitay commented:
@BrewTestBot test this please

fxcoudert commented on Formula/unit.rb:
```diff
@@ -0,0 +1,95 @@
+class Unit < Formula
```
Product page often calls it “NGINX Unit” so let's go for that

fxcoudert commented on Formula/unit.rb:
```diff
@@ -0,0 +1,95 @@
+  homepage "https://unit.nginx.org/"
+  url "https://unit.nginx.org/download/unit-1.1.tar.gz"
+  sha256 "7c66365f5ea87e7e8903bac9d639676947fe2ab7aa799ce35defc426b3409ee0"
+  head "https://hg.nginx.org/unit", :using => :hg
```
Remove head, which we only use for well-established formulas with large user base.

fxcoudert commented on Formula/unit.rb:
```diff
@@ -0,0 +1,95 @@
+  sha256 "7c66365f5ea87e7e8903bac9d639676947fe2ab7aa799ce35defc426b3409ee0"
+  head "https://hg.nginx.org/unit", :using => :hg
+
+  option "with-debug", "Build with debug logging enabled"
```
No option in new formulas. We are trying to make as many formulas as possible build without options, so users can benefit from our prebuilt binaries (bottles).

fxcoudert commented on Formula/unit.rb:
```diff
@@ -0,0 +1,95 @@
+  depends_on "openssl"
+  depends_on "pcre"
+
+  depends_on "go" => :optional
```
Same here: please choose a “default” set of languages, that are likely to be used by most users, and ship those by default.

sj26 replied:
This is trickier. Unit is designed to have shared objects for each runtime environment you want. A bit like php modules. And they are compiled with the same source tree as the daemon itself.

What if we had an nginx unit formula which came with shared objects linked against the base system's ruby, perl, python, then had an "nginx unit go" formula, an "nginx unit ruby" formula, etc, which link the nginx unit and go formulas, etc? This could be done without options, just more formulae.

fxcoudert replied:
Not sure how much these would be used…

fxcoudert commented on Formula/unit.rb:
```diff
@@ -0,0 +1,95 @@
+  depends_on "pcre"
+
+  depends_on "go" => :optional
+  depends_on "perl" => :optional
```
Can system Perl and Ruby be used here instead?

fxcoudert commented on Formula/unit.rb:
```diff
@@ -0,0 +1,95 @@
+      configure_args << "--debug"
+    end
+
+    system "./configure", *configure_args
```
No extra variable `configure_args`, inline all arguments to configure here.

sj26 replied:
I started that way, then extracted them when "--debug" was optional. Without that option, no wuckers. 👌

fxcoudert commented on Formula/unit.rb:
```diff
@@ -0,0 +1,95 @@
+
+    configure_args = [
+      "--prefix=#{prefix}",
+      "--bindir=#{bin}",
```
Once prefix is chosen, let `bin` and `sbin` be determined as upstream prefers.

sj26 replied:
This is again tricky. Unit is trying to put itself all into one directory, like an `/opt/unit`, which it also runs out of. I split it up a bit so it would link nicely into the `/usr/local` hierarchy, symmetric to the nginx formula and other unix-y daemon formulae, and make it easy to add shared objects in other formulae.

It's also a little tricky because most of these options are set at compile time, or on the command line — there is no config file.

A bare `./configure` (and perl, python, ruby) gives:

```
Configuration summary:

  unit bin directory:           "bin"
  unit sbin directory:          "sbin"
  unit modules directory:       "build"
  unit state directory:         "build"

  unit pid file:                "unit.pid"
  unit log file:                "unit.log"

  unit control API socket:      "unix:control.unit.sock"
```

and builds and installs into the current directory:

```
.
├── CHANGES
├── LICENSE
├── Makefile
├── NOTICE
├── README
├── auto
│   ├── atomic
│   ├── cc
│   └── ...
├── build
│   ├── Makefile
│   ├── autoconf.data
│   ├── autoconf.err
│   ├── ...
│   ├── perl.unit.so
│   ├── python.unit.so
│   ├── ruby.unit.so
│   ├── ...
│   ├── src
│   │   ├── nxt_app_log.dep
│   │   ├── nxt_app_log.o
│   │   ├── nxt_application.dep
│   │   ├── nxt_application.o
│   │   ├── ...
│   │   └── test
│   └── unitd
├── configure
├── sbin
│   └── unitd
├── src
│   ├── perl
│   │   └── ...
│   ├── nxt_aix_send_file.c
│   ├── nxt_app_log.c
│   ├── nxt_application.c
│   ├── nxt_application.h
│   ├── ...
│   └── test
│       ├── perl
│       │   └── ...
│       ├── nxt_gmtime_test.c
│       ├── nxt_http_parse_test.c
│       └── ...
└── test
    ├── perl
    │   └── ...
    ├── run.py
    ├── test_access_log.py
    └── ...

83 directories, 554 files
```

Just doing `./configure --prefix=...` (and perl, python, ruby) gives output like:

```
Configuration summary:

  unit bin directory:           "/usr/local/Cellar/nginx_unit/1.1/bin"
  unit sbin directory:          "/usr/local/Cellar/nginx_unit/1.1/sbin"
  unit modules directory:       "/usr/local/Cellar/nginx_unit/1.1/build"
  unit state directory:         "/usr/local/Cellar/nginx_unit/1.1/build"

  unit pid file:                "/usr/local/Cellar/nginx_unit/1.1/unit.pid"
  unit log file:                "/usr/local/Cellar/nginx_unit/1.1/unit.log"

  unit control API socket:      "unix:/usr/local/Cellar/nginx_unit/1.1/control.unit.sock"
```

and installs as a tree like:

```
/usr/local/Cellar/nginx_unit/1.1
├── build
│   ├── perl.unit.so
│   ├── python.unit.so
│   └── ruby.unit.so
├── sbin
│   └── unitd
├── unit.log
└── unit.pid
```

with all of its runtime stuff created within. There's nowhere to install shared modules from other formulae, logs are in unexpected places, etc.

Installing per the proposed formula give configure output:

```
Configuration summary:

  unit bin directory:           "/usr/local/Cellar/nginx_unit/1.1/bin"
  unit sbin directory:          "/usr/local/Cellar/nginx_unit/1.1/bin"
  unit modules directory:       "/usr/local/Cellar/nginx_unit/1.1/lib/unit"
  unit state directory:         "/usr/local/var/unit"

  unit pid file:                "/usr/local/var/run/unit.pid"
  unit log file:                "/usr/local/var/log/unit.log"

  unit control API socket:      "unix:/usr/local/var/run/unit.control.sock"

  non-privileged user:          "nobody"
  non-privileged group:         ""

  IPv6 support:                 YES
  Unix domain sockets support:  YES
  debug logging:                NO
```

and installs as a tree like:

```
├── bin
│   └── unitd
└── lib
    └── unit
        ├── perl.unit.so
        ├── python.unit.so
        └── ruby.unit.so
```

I'll see if I can get it to install the modules into `/usr/local/Cellar/nginx_unit/1.1/lib/unit` but load them from `/usr/local/lib/unit` for the multi-formulae approach.

sj26 replied:
The proposed layout here is pretty close to their debian/rpm/docker packing, too:

https://github.com/nginx/unit/tree/master/pkg

fxcoudert commented on Formula/unit.rb:
```diff
@@ -0,0 +1,95 @@
+      "--prefix=#{prefix}",
+      "--bindir=#{bin}",
+      "--sbindir=#{bin}",
+      "--modules=#{lib}/unit",
```
What would be the default value of `modules`, `state`,  `log`, `pid` and `control`?

sj26 replied:
Covered above.

fxcoudert commented on Formula/unit.rb:
```diff
@@ -0,0 +1,95 @@
+  end
+
+  test do
+    system "#{bin}/unitd", "--version"
```
We need a test that exercises the some of the functionality of the app.

sj26 replied:
I've added tests making sure the right files are in the right places, that the daemon is running and responding on the control port, and that the modules are being discovered (from the logs). I'll add some tests creating and running perl/plack, python/wsgi, and ruby/rack apps using the system runtimes tomorrow.

sj26 replied:
I've added some more tests making sure that each built module against the system frameworks — perl, python, and ruby.

DomT4 commented on Formula/nginx-unit.rb:
```diff
@@ -0,0 +1,100 @@
+        <false/>
+        <key>ProgramArguments</key>
+        <array>
+            <string>#{bin}/unitd</string>
```
`opt_bin`

DomT4 commented on Formula/nginx-unit.rb:
```diff
@@ -0,0 +1,267 @@
+      # This tells the unit daemon where it can load modules from. Configuring
+      # the location at runtime is painful. We want this to be in the
+      # $HOMEBREW_PREFIX so that other formulae can link in additional modules. 
+      "--modules=#{HOMEBREW_PREFIX/"lib/unit"}",
```
It's a nit but writing it this way is much nicer to look at:

```diff
diff --git a/Formula/nginx-unit.rb b/Formula/nginx-unit.rb
index b43c12a700..5a51c4fe2d 100644
--- a/Formula/nginx-unit.rb
+++ b/Formula/nginx-unit.rb
@@ -18,11 +18,11 @@ class NginxUnit < Formula
       # This tells the unit daemon where it can load modules from. Configuring
       # the location at runtime is painful. We want this to be in the
       # $HOMEBREW_PREFIX so that other formulae can link in additional modules.
-      "--modules=#{HOMEBREW_PREFIX/"lib/unit"}",
-      "--state=#{var/"unit"}",
-      "--log=#{var/"log/unit.log"}",
-      "--pid=#{var/"run/unit.pid"}",
-      "--control=unix:#{var/"run/unit.control.sock"}",
+      "--modules=#{HOMEBREW_PREFIX}/lib/unit",
+      "--state=#{var}/unit",
+      "--log=#{var}/log/unit.log",
+      "--pid=#{var}/run/unit.pid",
+      "--control=unix:#{var}/run/unit.control.sock",
       "--cc-opt=-I#{pcre.opt_include} -I#{openssl.opt_include}",
       "--ld-opt=-L#{pcre.opt_lib} -L#{openssl.opt_lib}"
 
@@ -62,7 +62,7 @@ class NginxUnit < Formula
         curl --unix-socket #{var}/run/unit.control.sock http://localhost/
 
       Modules are available for system perl, python and ruby. The system php doesn't support nginx-unit. You can install additional modules into:
-        #{HOMEBREW_PREFIX/"lib/unit"}
+        #{HOMEBREW_PREFIX}/lib/unit
     EOS
   end
```

sj26 replied:
I had it that way before, will switch it back.

DomT4 commented on Formula/nginx-unit.rb:
```diff
@@ -0,0 +1,267 @@
+    libunit = lib/"unit"
+    libunit.mkpath
+
+    # There is no system Go
```
Either use a `go` dependency or drop these lines.

DomT4 commented on Formula/nginx-unit.rb:
```diff
@@ -0,0 +1,267 @@
+    system "make", "perl"
+    libunit.install "build/perl.unit.so"
+
+    # We can't build against the system PHP, it doesn't have the right SAPI
```
Remove these lines.

sj26 replied:
It's worth leaving a note, it might seem like the php module should be installed if macOS has php, but it's in the caveats, too.

DomT4 commented on Formula/nginx-unit.rb:
```diff
@@ -0,0 +1,267 @@
+      Once running, you can control unit using the control socket, for example using curl:
+        curl --unix-socket #{var}/run/unit.control.sock http://localhost/
+
+      Modules are available for system perl, python and ruby. The system php doesn't support nginx-unit. You can install additional modules into:
```
Wrap this to 80.

DomT4 commented on Formula/nginx-unit.rb:
```diff
@@ -0,0 +1,267 @@
+  end
+
+  test do
+    system "#{bin}/unitd", "--version"
```
Don't need this.

sj26 replied:
Removed.

DomT4 commented on Formula/nginx-unit.rb:
```diff
@@ -0,0 +1,267 @@
+    require "timeout"
+
+    logfile = var/"log/unit.log"
+    logfile.unlink if logfile.exist?
```
I think this is permitted by the sandbox, _think_, but it'd be nicer to keep all the filesystem operations inside the temporary directory if possible.

sj26 replied:
It does seem to work (in the sandbox, during my tests) but was mostly because my log file kept growing when re-running the tests. I'll remove it.

DomT4 commented on Formula/nginx-unit.rb:
```diff
@@ -0,0 +1,267 @@
+      # temporary gems path before booting unitd. (unitd boots and functions
+      # fine without, but fails to start an app until rack is installed.)
+      rubydir = testpath/"ruby"
+      rubygems = rubydir/"gems"
```
Just do `rubygems = testpath/"ruby/gems"` here.

DomT4 commented on Formula/nginx-unit.rb:
```diff
@@ -0,0 +1,267 @@
+      rubydir = testpath/"ruby"
+      rubygems = rubydir/"gems"
+      rubygems.mkpath
+      ENV["GEM_PATH"] = rubygems.to_s
```
Is the `.to_s` required?

sj26 replied:
Apparently not! TIL. I'm just in the habit of being careful with ENV when writing compat ruby. I'll remove the to_s.

DomT4 commented on Formula/nginx-unit.rb:
```diff
@@ -0,0 +1,267 @@
+
+      pid = spawn "#{bin}/unitd", "--no-daemon"
+
+      timeout(5) { sleep 0.1 until pidfile.exist? }
```
This feels a bit... hacky. How unlikely is it the pidfile wouldn't exist in this scenario?

sj26 replied:
It might take a moment for unitd to boot and write the pidfile... depending on the speed of the system performing the test. I didn't want to leave a race condition that might become flakey. I did have a `sleep 2` like some other formulae but I got tired of waiting while re-running the tests. I can remove, or revert to `sleep 2`, or something, if you prefer?

DomT4 commented on Formula/nginx-unit.rb:
```diff
@@ -0,0 +1,267 @@
+      expected_body = {"listeners" => {}, "applications" => {}}
+      assert_equal expected_body, JSON.parse(response.body)
+
+      assert logfile.size > 0
```
You'd want some kind of failure message here or it'll potentially blow up in a way that'll be completely and utterly unhelpful to anyone.

sj26 replied:
```
````

seems to fail nicely:

```

DomT4 commented on Formula/nginx-unit.rb:
```diff
@@ -0,0 +1,267 @@
+
+      # Unit should discover each of the modules and talk about them in the log
+      logfile.read.tap do |logs|
+        #assert_include logs, "lib/unit/go.unit.so"
```
Cut any lines that aren't going to be created/installed.

DomT4 commented on Formula/nginx-unit.rb:
```diff
@@ -0,0 +1,267 @@
+
+      perldir = testpath/"perl"
+      perldir.mkpath
+      perlapp = perldir/"app.psgi"
```
`perlapp = testpath/"perl/app.psgi"`.

DomT4 commented on Formula/nginx-unit.rb:
```diff
@@ -0,0 +1,267 @@
+      perldir.mkpath
+      perlapp = perldir/"app.psgi"
+
+      perlapp.write <<~PERL
```
Standardise on `EOS`, but you can drop the `mkpath` above as the `.write` here will create the file as necessary.

sj26 replied:
I have syntax highlighting for heredocs based no the tag name, which is nice, but I'll change them to EOS.

I didn't know `#write` did `#mkpath` — neat! TIL.

DomT4 commented on Formula/nginx-unit.rb:
```diff
@@ -0,0 +1,267 @@
+      response = Net::HTTPResponse.read_new(socket)
+      response.reading_body(socket, request.response_body_permitted?) { }
+
+      pythondir = testpath/"python"
```
Same comments as from the above block really.

DomT4 commented on Formula/nginx-unit.rb:
```diff
@@ -0,0 +1,267 @@
+      rubydir.mkpath
+      rubyapp = rubydir/"config.ru"
+
+      rubyapp.write <<~RUBY
```
And same comments from above again.

sj26 replied:
It's difficult when it's WET 😅

sj26 commented:
Thanks for the reviews, @fxcoudert and @DomT4! I've addressed your feedback in an additional commit to make it easier to review, I'll squash when you're ready.

fxcoudert commented:
@ilovezfs any opinion here? on the test, in particular

fxcoudert commented:
@sj26 test seems to be failing on CI (time out)

sj26 commented:
Does CI have different sandbox rules? Afaict it should be working... but it’s not finding the pid or control socket, despite the service claiming to be started.

ilovezfs commented on Formula/nginx-unit.rb:
```diff
@@ -0,0 +1,256 @@
+
+    system "./configure",
+      "--prefix=#{prefix}",
+      "--bindir=#{bin}",
```
Is this not the default?

sj26 replied:
Oh, yes it is. Will remove.

ilovezfs commented on Formula/nginx-unit.rb:
```diff
@@ -0,0 +1,256 @@
+    system "./configure",
+      "--prefix=#{prefix}",
+      "--bindir=#{bin}",
+      "--sbindir=#{bin}",
```
Why are we relocating things out of sbin? Does it by default installing some things in bin and others in sbin?

sj26 replied:
It doens't put anything into bin by default, only `unitd` into sbin. But there is no sbin in `/usr/local`, and macOS doesn't have it in $PATH. So I followed the example from the nginx formula and used bin.

See the design behind these paths in an earlier comment:

https://github.com/Homebrew/homebrew-core/pull/27260#discussion_r190568505

ilovezfs replied:
I'd leave it as sbin.

sj26 replied:
This now fails for me:

```
homebrew-core (unit*) % brew reinstall nginx-unit
==> Reinstalling nginx-unit
==> Downloading https://unit.nginx.org/download/unit-1.1.tar.gz
Already downloaded: /Users/sj26/Library/Caches/Homebrew/nginx-unit-1.1.tar.gz
==> ./configure --prefix=/usr/local/Cellar/nginx-unit/1.1 --modules=/usr/local/lib/unit --state=/usr/
==> make install
==> ./configure perl
==> make perl
==> ./configure python
==> make python
==> ./configure ruby
==> make ruby
Error: The `brew link` step did not complete successfully
The formula built, but is not symlinked into /usr/local
Could not symlink sbin/unitd
/usr/local/sbin is not writable.

You can try again using:
  brew link nginx-unit
==> Caveats
Once running, you can control unit using the control socket, for example using
curl:
  curl --unix-socket /usr/local/var/run/unit.control.sock http://localhost/

Modules are available for system perl, python and ruby. The system php doesn't
support nginx-unit. You can install additional modules into:
  /usr/local/lib/unit

To have launchd start nginx-unit now and restart at login:
  brew services start nginx-unit
Or, if you don't want/need a background service you can just run:
  /usr/local/sbin/unitd --no-daemon
==> Summary
🍺  /usr/local/Cellar/nginx-unit/1.1: 11 files, 384.1KB, built in 31 seconds
```

ilovezfs replied:
what's the output of
```
ls -ld /usr/local/sbin
```

sj26 replied:
```
$ ls -ld /usr/local/sbin
ls: /usr/local/sbin: No such file or directory
```

```
$ mkdir /usr/local/sbin
mkdir: /usr/local/sbin: Permission denied
```

Homebrew doesn't make an sbin during installation, and can't without sudo because of SIP, afaik, and doesn't offer the option to do it.

ilovezfs replied:
We do indeed make that directory https://github.com/Homebrew/install/blob/master/install#L206

you'll need to fix it locally :)

sj26 replied:
Wow! Mea culpa, sorry, I don't know why I thought otherwise. I wonder what happened to mine! :+1:

ilovezfs commented on Formula/nginx-unit.rb:
```diff
@@ -0,0 +1,256 @@
+      response = Net::HTTP.get_response(URI.parse("http://localhost:8000"))
+      assert_equal "200", response.code
+      assert_equal "Hello, ruby world!", response.body
+    rescue
```
We shouldn't really be rescuing in a test block.

sj26 replied:
It's impossible to debug the tests without the log output. Is there another way you recommend capturing log info for debugging tests?

ilovezfs replied:
Add a `raise "here"` to the test. And then run `brew test --debug`

sj26 replied:
That doesn't work for CI, though?

ilovezfs replied:
If it fails on CI, it needs to be debugged locally.

sj26 replied:
It passes locally. It only fails on CI.

ilovezfs replied:
OK, feel free to debug that with the rescue but let's make sure it doesn't get merged with that in there.

ilovezfs commented on Formula/nginx-unit.rb:
```diff
@@ -0,0 +1,256 @@
+      "--log=#{var}/log/unit.log",
+      "--pid=#{var}/run/unit.pid",
+      "--control=unix:#{var}/run/unit.control.sock",
+      "--cc-opt=-I#{pcre.opt_include} -I#{openssl.opt_include}",
```
There aren't configure options for openssl and pcre specifically?

sj26 replied:
No. Their [configure scripts](https://github.com/nginx/unit/blob/master/configure) are completely custom:

```
unit-1.1$ ./configure --help

./configure options:

  --cc=FILE            set C compiler filename, default: "clang"
  --cc-opt=OPTIONS     set additional C compiler options
  --ld-opt=OPTIONS     set additional linker options

  --prefix=DIRECTORY   set prefix for relative pathnames, default: none
  --bindir=DIRECTORY   set user executables directory name
                       default: "bin"
  --sbindir=DIRECTORY  set system admin executables directory name
                       default: "sbin"
  --modules=DIRECTORY  set modules directory name, default: "build"
  --state=DIRECTORY    set state directory name, default: "build"

  --pid=FILE           set pid filename, default: "unit.pid"
  --log=FILE           set log filename, default: "unit.log"

  --control=ADDRESS    set address of control API socket
                       default: "unix:control.unit.sock"

  --user=USER          set non-privileged processes to run as specified user
                       default: "nobody"
  --group=GROUP        set non-privileged processes to run as specified group
                       default: user's primary group

  --no-ipv6            disable IPv6 support
  --no-unix-sockets    disable Unix domain sockets support

  --debug              enable debug logging


  python OPTIONS       configure Python module
                       run "./configure python --help" to see available options

  php OPTIONS          configure PHP module
                       run "./configure php --help" to see available options

  go OPTIONS           configure Go module
                       run "./configure go --help" to see available options

  perl OPTIONS         configure Perl module
                       run "./configure perl --help" to see available options

  ruby OPTIONS         configure Ruby module
                       run "./configure ruby --help" to see available options
```

ilovezfs commented on Formula/nginx-unit.rb:
```diff
@@ -0,0 +1,256 @@
+    EOS
+  end
+
+  plist_options :manual => "unitd"
```
why not --no-daemon here too?

sj26 replied:
Good point!

sj26 replied:
e.g. https://github.com/Homebrew/homebrew-core/blob/a06f1128a44ed57549bd8ea55d98755491d203c2/Formula/nginx.rb#L146 vs https://github.com/Homebrew/homebrew-core/blob/a06f1128a44ed57549bd8ea55d98755491d203c2/Formula/nginx.rb#L161-L163

ilovezfs commented on Formula/nginx-unit.rb:
```diff
@@ -0,0 +1,256 @@
+      curl:
+        curl --unix-socket #{var}/run/unit.control.sock http://localhost/
+
+      Modules are available for system perl, python and ruby. The system php doesn't
```
The bit about php seems irrelevant.

sj26 replied:
Folks installing unit might expect it to be able to run php apps using the macOS-installed php. It seems an important caveat?

ilovezfs replied:
Can we build the php module if we use Homebrew's php?

sj26 replied:
I tried, it's difficult. This much seemed like a good first step, building against everything available on the system without adding dependencies. I was going to propose an `nginx-unit-php` formula separately which builds just the php shared module against homebrew php and puts in into the `/usr/local/lib/unit` directory. Potentially also `nginx-unit-php@5.6` etc, as many versions can be used simultaneously — part of the pitch of unit. This could be done for other homebrew formulae with interop, like homebrew' perl, python, ruby, and go.

ilovezfs replied:
I see. In any case, I think the php reference should be removed since the fact that it doesn't work with macOS's php isn't specific to Homebrew and the caveats are for Homebrew specific information and path locations only.

ilovezfs commented on Formula/nginx-unit.rb:
```diff
@@ -0,0 +1,256 @@
+    openssl = Formula["openssl"]
+    pcre = Formula["pcre"]
+
+    system "./configure",
```
the whitespace should be aligned as
```
    system "./configure", "--prefix=#{prefix}",
                          "--modules=#{HOMEBREW_PREFIX}/lib/unit",
                          "--state=#{var}/unit",
```
etc.

sj26 commented:
Phew, tests are almost passing. Seems the tests can't reach out of testpath _at all_ on CI, but they can locally.

Last two are compatibility issues with el capitan and sierra ruby version:

```
==> gem install rack --install-dir=/tmp/nginx-unit-test-20180527-4265-1oxmdj4/ruby/gems
ERROR:  Error installing rack:
	rack requires Ruby version >= 2.2.2.
Error: nginx-unit: failed
```

I'll try pinning rack.

ilovezfs commented:
>Seems the tests can't reach out of testpath at all on CI, but they can locally.

There's no difference. Probably a file needs to be touched in post install that already exists locally

ilovezfs commented on Formula/nginx-unit.rb:
```diff
@@ -0,0 +1,223 @@
+    (var/"run").mkpath
+  end
+
+  def caveats
```
```
def caveats; <<~EOS
```

sj26 replied:
Done!

ilovezfs commented on Formula/nginx-unit.rb:
```diff
@@ -0,0 +1,223 @@
+    system "make", "install"
+
+    libunit = lib/"unit"
+    libunit.mkpath
```
this line isn't needed since you use `.install` below which does a `mkdir -p` first

sj26 replied:
Neat!

ilovezfs commented on Formula/nginx-unit.rb:
```diff
@@ -0,0 +1,223 @@
+
+    system "make", "install"
+
+    libunit = lib/"unit"
```
I'd just make it
```
    (lib/"unit").install "build/perl.unit.so", "build/python.unit.so",
                         "build/ruby.unit.so"
```
at the very end. Or does it delete them in between?

sj26 replied:
It does not — done!

ilovezfs commented on Formula/nginx-unit.rb:
```diff
@@ -0,0 +1,223 @@
+      request = Net::HTTP::Put.new("/config")
+      request.body = JSON.generate(
+        "listeners" => { "*:8000" => { "application" => "perlapp" } },
+        "applications" => { "perlapp" => { "type" => "perl", "working_directory" => perlapp.parent, "script" => perlapp } },
```
tweak the formatting to avoid going beyond the 80th column please, here and below

sj26 replied:
Done!

ilovezfs commented on Formula/nginx-unit.rb:
```diff
@@ -0,0 +1,223 @@
+
+      control = testpath/"unit.control.sock"
+
+      pid = spawn "#{sbin}/unitd", "--no-daemon", "--log", "/dev/stdout", "--control", "unix:#{control}"
```
wrap to avoid going beyond the 80th column

sj26 replied:
Done!

ged commented:
I am still interested in seeing this merged, stalebot.

ilovezfs commented:
that'll hush 🤖 up for good

zbeekman commented:
@sj26 When you have a chance it would be great to address @ilovezfs' remaining comments to get this PR over the finish line. Thanks!

leonklingele commented:
Should be bumped to v1.3 before being merged.

leonklingele commented on Formula/nginx-unit.rb:
```diff
@@ -0,0 +1,223 @@
+class NginxUnit < Formula
+  desc "Dynamic web and application server for multiple languages"
+  homepage "https://unit.nginx.org/"
+  url "https://unit.nginx.org/download/unit-1.1.tar.gz"
```
Should be bumped to v1.3 before being merged.

javian commented:
PIng @sj26 ^^

fxcoudert commented:
Closing for now, we'll be glad to reopen or consider an updated PR addressing the rest of the issues raised!

sj26 commented:
Sorry, I got a bit busy, but I've come back and updated the formula, and addressed @ilovezfs' feedback. Could we reooen?

sj26 commented:
It doesn't seem to have picked up my new commits, here:

https://github.com/Homebrew/homebrew-core/compare/sj26:unit
